User info (backend and frontend part)

RazSKg opened the pull request.

RazSKg committed:
add server side of getting user info (without authorization), fixed typhos, data-time json normalization

RazSKg committed:
add front-end part of viewing User info

add components: UserAvatar, UserInfo, UserAvatarWithInfo (whitch agrigate 2 previous)

RazSKg committed:
Align components in center for "user info page"

RazSKg committed:
add redirection from Profile to SignUp page if user is inauthorized, add docstrings.

serge30 commented on easyrest/models/user.py:
```diff
@@ -0,0 +1,29 @@
+    email = Column(Text)
+    phone_number = Column(Text)
+    birth_date = Column(Date)
+    role_id = Column(Integer)  # now it is without foreign key relations
```
Common practice to use special todo comments for such things:
```
# TODO: Add relations
```

serge30 commented on easyrest/scripts/json_helpers.py:
```diff
@@ -1,7 +1,17 @@
+import datetime
+
+
+def date_time_normalize(date_time_object):
```
Docstring.

serge30 commented on easyrest/views/user_info_controller.py:
```diff
@@ -0,0 +1,47 @@
+    # but now is simple User with Id = 1
+    user_id = 1
+    user = request.dbsession.query(User).get(user_id)
+    # debug (will replace when authorization will be done
```
Please use TODO comments.

serge30 commented on easyrest/views/user_info_controller.py:
```diff
@@ -0,0 +1,47 @@
+        raise HTTPForbidden("403")
+    else:
+        user_dict = user.as_dict()
+        user_dict = {key: date_time_normalize(value) for key, value in user_dict.items()}
```
In pull #8  Max fixes `as_dict` method to handle datetimes, so probably you will not need this.

RazSKg replied:
Yes, i saw it. if you don't mind - I will remove this after Max will merge his branch with "develop". Then I will use his mechanism. Ok?

serge30 replied:
Ok

RazSKg committed:
added TODO comments, docstring.

RazSKg committed:
on frontend add function which redirect to SignUp page when forbidden, edit TODO comment

RazSKg commented:
also added function, which redirect fo SignUp page when Forbidden.

viktormovchaniuk commented on easyrest/views/user_info_controller.py:
```diff
@@ -0,0 +1,47 @@
+    else:
+        user_dict = user.as_dict()
+        user_dict = {key: date_time_normalize(value) for key, value in user_dict.items()}
+        body = wrap([user_dict])
```
look like you don't need wrap your userdict into brackets

RazSKg replied:
but wrap function data paramether is "data (list): list of dictionaries representing models output". I must do that.

serge30 replied:
I believe, description to this param is misleading. It doesn't require list here. Please send a record instead of list, and fix doctsring for `wrap` function.

viktormovchaniuk commented on frontend/src/components/UserInfo/UserAvatarWithInfo.js:
```diff
@@ -0,0 +1,46 @@
+
+const styles = {
+    forDiv: {
+        flexGrow: 1,
```
Don't need `flex-grow` property here. By default flex container stretch to fill the space

RazSKg replied:
Ok. Thanks!

viktormovchaniuk commented on frontend/src/components/UserInfo/UserAvatar.js:
```diff
@@ -0,0 +1,54 @@
+        width: "50%",
+        float: "left",
+        margin: 16,
+
```
You don't really need `float` property almost in all cases, however your `float` isn't work here because your element inside `flex` container
Please for positioning  your elements use `<Grid/>` component or at least `flex-box` model

RazSKg replied:
Thanks for info. I will use Grid.

viktormovchaniuk commented on frontend/src/components/UserInfo/UserAvatarWithInfo.js:
```diff
@@ -0,0 +1,46 @@
+    render() {
+        const {classes} = this.props;
+        const {userInfo, isForbidden} = this.state;
+        console.log(isForbidden);
```
don't commit debugging code

RazSKg replied:
all logs I removed (but didn't commit yet) - I forgot remove them before commit.

viktormovchaniuk commented on frontend/src/components/UserInfo/UserInfo.js:
```diff
@@ -0,0 +1,73 @@
+    },
+    cardStyle: {
+        width: "100%",
+        float: "left",
```
floats

RazSKg replied:
fixed.

viktormovchaniuk commented on frontend/src/components/UserInfo/UserAvatarWithInfo.js:
```diff
@@ -0,0 +1,46 @@
+        const {classes} = this.props;
+        const {userInfo, isForbidden} = this.state;
+        console.log(isForbidden);
+        if (isForbidden){
```
as i understood it is temporary, right?

RazSKg replied:
all logs I removed (but didn't commit yet) - I forgot remove them before commit.

viktormovchaniuk replied:
i mean isForbidden cheeking, not console.log()

RazSKg replied:
If isForbidden is "true" - I return Redirection, else I return false. It influence in "UserAvatarWithInfo.js" in render() method

RazSKg replied:
Sorry, this is outdated code. Those changes, which you describe - wasn`t presented in yesterday`s commit.

RazSKg committed:
remove float, flex property

serge30 commented on easyrest/scripts/insert_example_data.py:
```diff
@@ -81,5 +83,19 @@ def fill_db(session):
         Rest_models.append(rest_model)
 
+    # add users
+    for myIter in range(menu_item_number):
```
PEP8 for variable name. Anyway `iter` is bad suffix here as iterator is object we may iterate, instead of particular record. I'd use just `i` or `index` here.

serge30 commented on frontend/src/components/UserInfo/UserAvatarWithInfo.js:
```diff
@@ -0,0 +1,61 @@
+        fetch('http://localhost:6543/profile', {method: "GET", headers})
+            .then(response => response.json())
+            .then(data => this.setState({userInfo: data.data[0], success: data.success, error: data.error}))
+            .catch(err=>console.log(err))
```
Most of users don't know nothing about browser console. You need to let them now, that something went wrong.

serge30 commented on frontend/src/Service/NeedAuthorization.js:
```diff
@@ -0,0 +1,20 @@
+    /*this function checks error message for Forbidden cause.
+    If it present - return Redirect for LogIn page, else - return false*/
+
+    const isForbidden = ("" + error).toLowerCase().search("forbidden") !== -1;
```
I'm not sure about this approach, it's unsafe as you cannot be sure that other errors will not contain "forbidden" word. You'd better add some field to response dict along with `success` and `error` which indicate unauthorized access, or use proper HTTP status code 403.

RazSKg committed:
backend: set proper http status for 404 and 403, minor changes. frontend: rework redirection, add error page.

Backend: In 403 and 404 views set proper http status code. Make corrections according to
pull request remarks.
Frontend:
Make redirection in case of 403 status code in http header. Add pake for errors.

RazSKg commented:
Made corrections according to Your remarks. Add http status code for NotFound and Forbidden. Made redirection base on http status code. Add page for errors (instead console.log()).

serge30 reviewed (commented):
@RazSKg Please resolve conflicts.

RazSKg committed:
add view and route to handle OPTION HTTP request, remove interactive Card component

RazSKg committed:
Merge branch 'develop' into user_view_profile. Resolving conflicts.

MaxSukhovetskyi committed:
Fixed datetime to string, added profile button in user menu, fixed bug with no my restaurants in user menu

RazSKg commented:
@serge30 ready for review

serge30 commented on frontend/src/components/UserInfo/UserAvatarWithInfo.js:
```diff
@@ -0,0 +1,66 @@
+            return redirection;
+        }
+        else{
+            if(success ===true){
```
But why? What's wrong with `if(success) {...`?

serge30 commented on frontend/src/components/UserMenu.jsx:
```diff
@@ -65,7 +65,8 @@ class UserMenu extends React.Component {
 
     const { auth, token, role } = this.props.ctx;
-    const isOwner = role === "Owner";
+    console.log(role)
```
No debug loging in code please.

serge30 commented on easyrest/scripts/json_helpers.py:
```diff
@@ -1,13 +1,30 @@
+import datetime
+
+
+def date_time_normalize(date_time_object):
```
I do not think that `date_time_normalize` is still used.

serge30 commented on easyrest/models/__init__.py:
```diff
@@ -17,6 +17,7 @@
 from .tag import Tag
 from .tag_association import TagAssociation
 from .user import User
+
```
What is the purpose of this change?

RazSKg replied:
It`s about space? If yes - I accidentally added this.

serge30 reviewed (approved):
Please make a fixes according to comments and merge.

RazSKg committed:
fixed pull request remarks.
